session_pool.go

```diff
@@ -384,6 +394,50 @@ func (pool *SessionPool) getIdleSession() (*Session, error) {
 		" session pool and the total session count has reached the limit")
 }
 
+// retryGetSession tries to create a new session when the current session is invalid.
+func (pool *SessionPool) executeWithRetry(
+	session *Session,
+	f func(*Session) (*graph.ExecutionResponse, error),
+	retry int) (*graph.ExecutionResponse, error) {
+	pool.rwLock.Lock()
+	defer pool.rwLock.Unlock()
```

Comment:
Should it be possible to reduce the granularity of locks?
Only need to lock *list.List, not during session creation, execution, etc.

Reply:
Yeah, we should recheck the lock usage later.